### PR TITLE
nix-flake: Add shell definition

### PR DESCRIPTION
**Pull request by markoburcul**
Initialized flake file and added development shell definition.

Referenced issue: https://github.com/codex-storage/nim-codex/issues/940

### CODE REVIEW

**Comment by markoburcul** [comment]
The shell example taken from @corpetty [comment](https://github.com/codex-storage/nim-codex/issues/940#issuecomment-2397079978). I've managed to reuse this shell.nix file with minimal modifications. To test the shell I have:
- built codex executable
- build/test and launch Hardhat node
- run `make testAll` with success
- run the [`Local two client test`](https://docs.codex.storage/learn/local-two-client-test) with success

Now, I am not 100% sure about this part:

> Here's how I updated the CMakeLists.txt file (from Claude.ai):
> 
>     1. Open the CMakeLists.txt file in a text editor:
>        `nano /home/petty/codex/repos/nim-codex/vendor/nim-leopard/vendor/leopard/CMakeLists.txt`

since we do set `CXXFLAGS` and `CFLAGS` for the shell and for the derivation in the `flake.nix` file you've sent. Maybe I didn't understand well..

File: flake.nix [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    flake-utils.url = "github:numtide/flake-utils";
+  };
+
+  outputs = { self, nixpkgs, flake-utils }:
+    flake-utils.lib.eachDefaultSystem (system:
+      let
+        pkgs = nixpkgs.legacyPackages.${system};
+
+        # Create a dummy lsb_release script
+        dummy-lsb-release = pkgs.writeScriptBin "lsb_release" ''
```
Comment (yakimant) [inline flake.nix:15]: Is it related?
https://github.com/NixOS/nixpkgs/tree/master/pkgs/os-specific/linux/lsb-release

**Review by yakimant — APPROVED** [review]
Looks good.
Althtough, I would add `.envrc` for nix flakes every time we add `flake.nix` - it's so fast and convinient!

File: flake.nix (hunk 1 above)
Comment (jakubgs) [reply flake.nix:15]: Yes, I think we can use `lsb-release` from `nixpkgs` itself.

File: flake.nix [hunk 2]
```diff
@@ -0,0 +1,58 @@
+{
+  description = "Codex build flake";
+
+  inputs = {
+    nixpkgs.url = "github:NixOS/nixpkgs/nixos-unstable";
+    flake-utils.url = "github:numtide/flake-utils";
+  };
+
+  outputs = { self, nixpkgs, flake-utils }:
+    flake-utils.lib.eachDefaultSystem (system:
```
Comment (jakubgs) [inline flake.nix:10]: I think I prefer avoiding `flake-utils`, the same can be achieved with a few lines:
https://github.com/status-im/go-waku/blob/3c4e40c729a0ad5f1079ac1e3f31eeaae1b9cbdd/flake.nix#L8-L13

But I'm not saying you have to change it.

File: flake.nix [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+                bash
+                pkg-config
+                openssl
+                gmp
+                dummy-lsb-release
+                nim
+                nimble
+                nodejs_18
+                llvmPackages.openmp
+            ];
```
Comment (jakubgs) [inline flake.nix:39]: I don't think separating it all like that into multiple lines is great. I'd do something more organized:
```suggestion
            buildInputs = with pkgs; [
              # General
              git pkg-config openssl lsb-release
              # Build
              rustc cargo nim nimble gcc11 cmake
              # Libraries
              gmp llvmPackages.openmp
              # What is this for?
              nodejs_18
            ];
```

File: flake.nix [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+          #!${pkgs.stdenv.shell}
+          echo "Distributor ID: NixOS"
+          echo "Description:    NixOS"
+          echo "Release:        Unstable"
+          echo "Codename:       nixos"
+        '';
+      in {
+        devShell = pkgs.mkShell {
+            buildInputs = with pkgs; [
+                gcc11
```
Comment (jakubgs) [inline flake.nix:25]: I think the more kosher way of doing this would involve changing the GCC for the whole shell:
```nix
mkShell = mkShell.override { stdenv = pkgs.gcc11Stdenv; };
```
Something like that.

File: flake.nix [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+                nim
+                nimble
+                nodejs_18
+                llvmPackages.openmp
+            ];
+
+            shellHook = ''
+                export PATH=${pkgs.gcc11}/bin:${dummy-lsb-release}/bin:$PATH
+                export CC=${pkgs.gcc11}/bin/gcc
+                export CXX=${pkgs.gcc11}/bin/g++
```
Comment (jakubgs) [inline flake.nix:44]: Should not be necessary if you adjust `stdenv` used.

**Review by jakubgs — CHANGES_REQUESTED** [review]
Taking shape.

File: flake.nix (hunk 3 above)
Comment (markoburcul) [reply flake.nix:39]: The nodejs18 is used for running tests https://docs.codex.storage/learn/local-two-client-test . I assumed if it is on their website listed as official test that it would be neat to have that possibility in the shell.

File: flake.nix [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+              # Libraries
+              gmp llvmPackages.openmp
+              # Tests
+              nodejs_18
+          ];
+          shellHook = ''
+              export NIMBLE_DIR=$HOME/.nimble
+              export RUSTFLAGS="-C target-cpu=native"
+              export CXXFLAGS="-std=c++17 -march=native -mtune=native -msse -msse2 -msse3 -mssse3 -msse4.1 -msse4.2 -mavx -mavx2 -fopenmp"
+              export CFLAGS="-march=native -mtune=native -msse -msse2 -msse3 -mssse3 -msse4.1 -msse4.2 -mavx -mavx2"
```
Comment (jakubgs) [inline flake.nix:36]: Why do we need these? Were they recommended by devs? If it is it probably should be set in the `Makefile`.

File: flake.nix [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+              mkdir -p $NIMBLE_DIR
+              echo "Codex build environment loaded."
+              echo "GCC path: $(which gcc)"
+              echo "G++ path: $(which g++)"
```
Comment (jakubgs) [inline flake.nix:40]: What's the utility of this?

**Review by jakubgs — APPROVED** [review]
Tested locally, works well.

File: flake.nix (hunk 7 above)
Comment (markoburcul) [reply flake.nix:40]: Maybe just informative to the one who will be developing the code within the project

File: flake.nix (hunk 6 above)
Comment (markoburcul) [reply flake.nix:36]: I didn't ask(I will) but without them build would fail

Comment (jakubgs) [reply flake.nix:36]: That's weird. If it's required for builds it should be in `Makefile`. It would be good to know what part of this is actually necessary. I'd rather avoid [cargo culting](https://en.wikipedia.org/wiki/Cargo_cult) stuff into config files that are already hard to follow for most people.

Comment (markoburcul) [reply flake.nix]: I've moved the CXXFLAGS env variable to makefile with **minimum** number of compiler flags in order for build to work. I'll add someone from codex team so they can see this change and comment on it.

**Comment by markoburcul** [comment]
While testing the compiling with system nim compiler(`make USE_SYSTEM_NIM=1`), I realized the [nim](https://search.nixos.org/packages?channel=24.05&from=0&size=50&sort=relevance&type=packages&query=nim) from nixpkgs is version 2.0.4 and we can't compile the code with it. I also tried [nim1](https://search.nixos.org/packages?channel=24.05&from=0&size=50&sort=relevance&type=packages&query=nim) which is version 1.6.20 (much closer to the one we use in the Makefile which is a commit dating after official 1.6.20 release) and it failed. Only with [nim-unwrapped-1](https://search.nixos.org/packages?channel=24.05&from=0&size=50&sort=relevance&type=packages&query=nim) the compiling is successfull so I added it to the list of packages in the shell.

**Comment by gmega** [comment]
I'm going to merge this as the people who use Nix seem to be happy with it.

File: flake.nix (hunk 6 above)
Comment (gmega) [reply flake.nix]: I never had to set anything of the kind to build Codex, I wonder why this is needed here.

**Comment by gmega** [comment]
@markoburcul I'm probably missing a lot of context as to how this works, but our build system should build the Nim version that's required by Codex. You shouldn't, indeed, use an external compiler as that might not work. Can't we just lean on that?

**Comment by markoburcul** [comment]
> @markoburcul I'm probably missing a lot of context as to how this works and what the constraints are, but our build system should build the Nim version that's required by Codex. You shouldn't, indeed, use an external compiler as that might not work. Can't we just lean on that?

Using `nix shell` to build the code ensures consistency across environments by specifying **exact dependencies and versions**, including the Nim compiler. Adding a specific Nim compiler version in nix doesn’t force its use; it just makes that version available in the environment. You could use it by specifying `USE_SYSTEM_NIM=1` flag when running `make`. What I was more interested in was you're opinion on these CXXFLAGS I'm settng as default in the Makefile, because they are required to build `codex` without errors.

If you lack more context on this you can check out the task [here](https://github.com/codex-storage/nim-codex/issues/940).

**Comment by jakubgs** [comment]
@gmega are you using a very specific version of Nim compiler because you need very specific patches. Or did you just copy what `nimbus-eth2` had like a year ago or more than that without thinking much about it and now are just continuing with it?
